---
title: "Analysis of the RGC single-cell sequencing data"
author: "Miguel Fernandes"
date: "11/23/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load packages

We start by loading the packages required (use install.packages if needed).

```{r packages, message=FALSE}
library(dplyr)
library(Seurat)
library(Matrix)
library(ggplot2)
library(sctransform)
library(reticulate)
library(SeuratWrappers)
library(harmony)
library(reshape2)
```

### Remove immediate early genes (IEGs)
Based on Inbal Shanier scripts
IEGs expression can be elevated in certain cells as a response to the dissection and dissociation of the cells. This can hamper the analysis of the cell types and therefore we will exclude IEGs from the analysis (see this paper for additional information: <https://pubmed.ncbi.nlm.nih.gov/29024657/>). We have curated a list of zebrafish known IEGs, and we'll use this function to exclude them:  

```{r remove_IEGs}
remove_genes <- function(dgc_mat, gene_list) {
  
  #the function receives a dgc gene matrix and a list of genes and returns a dgc gene matrix without the gene rows. The loaded 10x data is a dgc matrix.
  
  ieg.indices = row.names(dgc_mat) %in% gene_list
  dgc_mat.mod = dgc_mat[!ieg.indices,]
  
  
  return (dgc_mat.mod)
  
}
```

Let's load our IEGs curated list:

```{r, load_IEGs_list}
IEGs<- readLines("~/desktop/ZebrafishRGC_data/IEG list.csv")
```

### Load the data

We will load our data from the 10x matrix files. Then we'll remove the IEGs, create a seurat object, and add additional metadata about the larvae age. We will do that for each of the scSeq batches. 

```{r load_data, message=FALSE, warning=FALSE}
# load the 10x data:
larva.7dpf.data<- readRDS("~/desktop/ZebrafishRGC_data/larva_zFish_FINAL.rds")

# Remove ZfishRGC17 due to wetting sample failure
cells.remove = grep("ZfishRGC17_",colnames(larva.7dpf.data), value=TRUE)
larva.7dpf.data = larva.7dpf.data[, setdiff(colnames(larva.7dpf.data), cells.remove)]


# remove the IEGs using the function and the list described above.
larva.7dpf.data.no.iegs <- remove_genes(larva.7dpf.data, IEGs)
# if needed to create a seurat object. rds files already have a Seurat object

#larva.no.iegs <- CreateSeuratObject(counts = larva.7dpf.data.no.iegs, project = "larvaRGC", min.cells = 25, min.features = 450)
# add metadata about the larvae age
larva.no.iegs<- AddMetaData(object=larva.7dpf.data.no.iegs, metadata = "7dpf", col.name="age")
```

Get the summary of the generated object. Features is the number of genes, samples is the number of cells.
```{r object_summary}
 
larva <-larva.no.iegs

```


```{r check_larva}
larva[["percent.mt"]] <- PercentageFeatureSet(larva, pattern = "^MT-")
larva[["percent.rps"]] <- PercentageFeatureSet(larva, pattern = "^RPS")
larva[["percent.rpl"]] <- PercentageFeatureSet(larva, pattern = "^RPL")
larva[["percent.rp"]] <- larva[["percent.rps"]] + larva[["percent.rpl"]]

# Create Violin Plots of RNA counts, mitochondrial
# percentages, and ribosomal percentages
VlnPlot(larva, features = "nCount_RNA", pt.size = 0.3)
```

```{r}
head(larva@meta.data)
```

Repeat for the adult data
```{r rest_of objects, message=FALSE, warning=FALSE}
# load the 10x data:
adult.7dpf.data<- readRDS("~/desktop/ZebrafishRGC_data/adult_zFish_FINAL.rds")

# Remove ZfishRGC17 due to wetting sample failure
cells.remove = grep("ZfishRGC17_",colnames(adult.7dpf.data), value=TRUE)
adult.7dpf.data = adult.7dpf.data[, setdiff(colnames(adult.7dpf.data), cells.remove)]


# remove the IEGs using the function and the list described above.
adult.7dpf.data.no.iegs <- remove_genes(adult.7dpf.data, IEGs)
# if needed to create a seurat object. rds files already have a Seurat object
#adult.no.iegs <- CreateSeuratObject(counts = adult.7dpf.data.no.iegs, project = "adultRGC", min.cells = 25, min.features = 450)
# add metadata about the adulte age
adult.no.iegs<- AddMetaData(object=adult.7dpf.data.no.iegs, metadata = "7dpf", col.name="age")
adult <-adult.no.iegs
```


```{r}
head(adult@meta.data)
```


Lets remove the unneeded objects from the environment to free memory. We will only need the merged seurat object from now on. 
```{r}
rm(larva.7dpf.data,
   larva.7dpf.data.no.iegs,
   larva.no.iegs,
   adult.7dpf.data,
   adult.7dpf.data.no.iegs,
   adult.no.iegs)
```




### Quality control

Removal of bad quality cells: 
1. Cell with high mitochondrial gene expression which usually sugests these are stressed cells. 
2. Cells with unusually high gene or UMI number, which usuaaly suggests there are douplets or triplets (originate from droplets containing more than one cell). 

We first calculate mitochondrial QC metrics with the PercentageFeatureSet function, 
which calculates the percentage of counts originating from a set of features. Mitochndrial genes have the prefix 'mt-', which is used to identify them.
```{r percent.mt}
larva[["percent.mt"]] <- PercentageFeatureSet(object = larva, pattern = "^mt-")
```

Lets visualize QC metrics as a violin plot (the features are the genes, the count is the number of UMIs).The visualization is according to the batch, which also help us to observe batch quality differences. 
```{r vln_plot}
VlnPlot(object = larva, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size=1)
```

We'll now filter out the cells with unusual number of genes (nFeature_RNA), precent.mt and UMI (nCount_RNA). What is unusual? No ground truth role here. It can vary with the tissue examined, the efficiency the 10x or the sequencing machine. We can examine the violin plot and get a feeling of the outliers in our data. Additionally for mitochondrial genes, the avg. in neurons is about 12.5%, so I would set the bat higher to capture most of the population variance. The most important part here is to check how the downstream analysis looks like, and return and refine the threshold set here. For example, if in the downstream analysis mitochondrial genes are highly present in the variable genes, the bar was set too high. If some cells have marker genes of two different cluster, maybe there are a lot od duplets in the data and the threshold for gene number and UMI should be lower. 

The subset command is used to keep the cells with certain characteristics:
```{r subset}
larva <- subset(x = larva, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 20 & nCount_RNA < 6000)
```

How many cells have past the threshold and will be used for the down stream analysis:
```{r}
sum(table(...=larva@active.ident))
```

### preprocessing

We will now normalize and scale the data. 
Normalize the data by a global-scaling normalization method. 
LogNormalize: Feature counts for each cell are divided by the total counts for that cell and multiplied by the scale.factor. This is then natural-log transformed using log1p.
More normalization methods can be used (CLR and Relative counts). Check the function help for more information. 

```{r normalization}
larva <- NormalizeData(object = larva, 
                                  normalization.method = "LogNormalize", 
                                  scale.factor = 10000)
```

Identification of highly variable features (feature selection). Identifies features that are outliers on a 'mean variability plot'.The default- 2,000 features per dataset. 
vst: First, fits a line to the relationship of log(variance) and log(mean) using local polynomial regression (loess). Then standardizes the feature values using the observed mean and expected variance (given by the fitted line). Feature variance is then calculated on the standardized values after clipping to a maximum (see clip.max parameter).

```{r variable features}
larva <- FindVariableFeatures(object = larva, 
                                         selection.method = "vst", 
                                         nfeatures = 2000)
```

Scaling the data:
Apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction. The ScaleData function:
Shifts the expression of each gene, so that the mean expression across cells is 0
Scales the expression of each gene, so that the variance across cells is 1
This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
The results of this are stored in combined.harmony[["RNA"]]@scale.data

```{r scaledata}
all.genes.no.iegs <- rownames(x = larva)
larva<- ScaleData(object =larva, features = all.genes.no.iegs)

```

### Principle component anaysis

Perform linear dimensional reduction (PCA on scaled data).
By default, only the previously determined variable features are used as input, 
but can be defined using features argument if you wish to choose a different subset.

```{r pca}
larva <- RunPCA(object = larva, features = VariableFeatures(object = larva))
```

### Batch correction

I use Harmony for batch correction. I have tried a couple other methods (Liger, fastMNN) and Harmony was superior to them. In their paper (<https://www.nature.com/articles/s41592-019-0619-0>) the describe the method as: "an algorithm that projects cells into a shared embedding in which cells group by cell type rather than dataset-specific conditions". The batches are refered by the "orig.ident", which we set in the add.cell.ids when generating the merged suerat object. 
```{r harmony}
larva <- RunHarmony(larva, group.by.vars = "orig.ident")
```

### Clustering and non-linear dimensional reduction
Explanation from the Seurat website:
"We first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical Mechanics], to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters function implements this procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents function".

The dims define the number of PCs used. The reduction takes the batch corrected data from Harmony.
```{r UMAP}
larva<- RunUMAP(larva, reduction = "harmony", dims = 1:30) 
```

```{r clsuters}
larva <- FindNeighbors(larva, reduction = "harmony", dims = 1:30) 
larva <- FindClusters(object = larva, resolution = 1.2)
```

Visualize the UMAP. First check the batch correction (defined by using the group.by):
```{r plot_umap_ident}
DimPlot(larva, group.by = c("orig.ident"))
```

Violin. First check the batch correction (defined by using the group.by):
```{r plot_umap_ident}
VlnPlot(larva,features = "nCount_RNA", group.by = c("orig.ident"))
```


And now the unique cell types:
```{r plot_umap_cluster}
DimPlot(object = (larva), label=TRUE)
```

### Finding differentially expressed features

Find markers for every cluster compared to all remaining cells. From seurat webpage:
"The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, max.cells.per.ident can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top".

To idetify the which kind of cells are in each cluster I define the setting to report only the positive markers, the marker has to be expressed in at least 25% of the cells in the cluster and to have a 
log fold expression of at least 0.8.

```{r find_features, message=FALSE, warning=FALSE}
larva.markers <- FindAllMarkers(object = larva, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.8)
```

Lets look on the top markers:
```{r markers}
larva.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
```

plot the expression of a chosen genes on the UMAP. elavl3 are differentiated neurons, while the others are progenitors markers:
```{r feature_plot}
FeaturePlot(object = larva, features = c("elavl3", "pcvna", "fabp7a", "her4.1"))
```

We can use this data to subset the differentiated neurons and recluster them. 
```{r subset_neurons}
larva<- subset(x = larva, idents=c("0", "1", "2", "3", "4","5", "10", "11", "14", "18", "22", "23"))
```

### Neurons analysis

how many neurons do we have:
```{r}
sum(table(...=larva@active.ident))
```

We will need to analyze the variable features for the new dataset and rerun the clustering analysis.
```{r analyze_neruons, message=FALSE, warning=FALSE}
# varibale features
larva <- FindVariableFeatures(object = larva, selection.method = "vst", nfeatures = 2000)
#pca 
larva <- RunPCA(object = larva, features = VariableFeatures(object = larva))
#Run harmony
larva <- RunHarmony(larva, group.by.vars = "orig.ident")
#RunUMAP (object, reduction = "harmony", dims = 1:30)
larva <- RunUMAP(larva, reduction = "harmony", dims = 1:30)
larva <- FindNeighbors(larva, reduction = "harmony", dims = 1:30) 
larva <- FindClusters(object = larva, resolution = 1.7)
```

Visualize the results. Firts batch effects:
```{r}
DimPlot(larva, group.by = c("orig.ident", "age"))
```

Clusters:

```{r}
DimPlot(object = (larva), label=TRUE)
```

Find markers for every neuronal cluster compared to all remaining cells, report only the positive ones.
```{r message=FALSE, warning=FALSE}
harmony.neurons.markers.no.iegs <- FindAllMarkers(object = larva, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.8)
```
```{r}
harmony.neurons.markers.no.iegs %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
```

Lets look on the neurotransmiters expression:
```{r}
FeaturePlot(object = larva,features = c("gad2", "gad1b", "slc6a1b", "chata", "slc17a6a", "slc17a6b", "slc17a7a", "slc6a5", "slc6a9"))
```

We can use this data to recluster only the inhibitory or the excitatory neurons etc. 

